flink-cdc-runtime/src/main/java/org/apache/flink/cdc/runtime/functions/SystemFunctionUtils.java

```diff
@@ -108,83 +108,203 @@ public static TimestampData toTimestamp(String str, String format, String timezo
+            default:
+                throw new RuntimeException(
+                        String.format(
+                                "Unsupported timestamp interval unit %s. Supported units are: SECOND, MINUTE, HOUR, DAY, MONTH, YEAR",
+                                timeIntervalUnit));
+        }
     }
 
-    public static int timestampDiff(
-            String symbol, TimestampData fromTimestamp, ZonedTimestampData toTimestamp) {
-        return timestampDiff(symbol, fromTimestamp.getMillisecond(), toTimestamp.getMillisecond());
+    public static LocalZonedTimestampData timestampadd(
+            String timeIntervalUnit,
+            Integer interval,
+            LocalZonedTimestampData timePoint,
+            String timezone) {
+        if (interval == null || timePoint == null) {
+            return null;
+        }
+        return LocalZonedTimestampData.fromEpochMillis(
+                timestampadd(
+                        timeIntervalUnit, interval, timePoint.getEpochMillisecond(), timezone));
     }
 
-    public static int timestampDiff(
-            String symbol, ZonedTimestampData fromTimestamp, TimestampData toTimestamp) {
-        return timestampDiff(symbol, fromTimestamp.getMillisecond(), toTimestamp.getMillisecond());
+    public static TimestampData timestampadd(
+            String timeIntervalUnit, Integer interval, TimestampData timePoint, String timezone) {
+        if (interval == null || timePoint == null) {
+            return null;
+        }
+        return TimestampData.fromMillis(
+                timestampadd(timeIntervalUnit, interval, timePoint.getMillisecond(), "UTC"));
     }
 
-    public static int timestampDiff(String symbol, long fromDate, long toDate) {
-        Calendar from = Calendar.getInstance();
-        from.setTime(new Date(fromDate));
-        Calendar to = Calendar.getInstance();
-        to.setTime(new Date(toDate));
-        Long second = (to.getTimeInMillis() - from.getTimeInMillis()) / 1000;
-        switch (symbol) {
+    private static long timestampadd(
+            String timeIntervalUnit, int interval, long timePoint, String timezone) {
+        Calendar calendar = Calendar.getInstance();
+        calendar.setTimeZone(TimeZone.getTimeZone(timezone));
+        calendar.setTime(new Date(timePoint));
+        int field;
+        switch (timeIntervalUnit) {
             case "SECOND":
-                return second.intValue();
+                field = Calendar.SECOND;
+                break;
             case "MINUTE":
-                return second.intValue() / 60;
+                field = Calendar.MINUTE;
+                break;
             case "HOUR":
-                return second.intValue() / 3600;
+                field = Calendar.HOUR;
+                break;
             case "DAY":
-                return second.intValue() / (24 * 3600);
+                field = Calendar.DAY_OF_YEAR;
+                break;
```

Comment:
Seems `DAY_OF_MONTH`, `DAY_OF_YEAR`, `DAY_OF_WEEK` are all synonyms of `Calendar.DATE` when calling `Calendar#add`. Maybe just use `DATE` for clarity?